ConwaysGameOfLife.Tests/GameControllerShould.cs

```diff
@@ -9,64 +10,121 @@ public class GameControllerShould
-            var universe = new Universe(gridWidth, gridLength);
-            var controller = new GameController(universe, _output, _input);
+            InputParser.ParseUniverse(universeInput);
             _input.GetReadLine(input);
             var expected = input;
 
-            var result = controller.CreateValidLocationString();
+            var result = _controller.CreateValidLocationString(gridWidth, gridLength);
 
             Assert.Equal(expected, result);
         }
 
-        //TODO: how to test error message in CreateValidLocationString
+        [Theory]
+        [InlineData(new string[] { "0,3" }, "2,2", "3,3", 3, 3)]
+        [InlineData(new string[] { "3,0" }, "0,0", "3,3" , 3, 3)]
+        [InlineData(new string[] { ",,", ";;" }, "2,2", "3,3", 3, 3)]
+        [InlineData(new string[] { "3,3,", "3,4;" }, "3,2", "4,3", 4, 3)]
+        [InlineData(new string[] { "3,3,3" },"3,2", "4,3", 4, 3)]
+        [InlineData(new string[] { "a,b" }, "1,1", "3,3", 3, 3)]
+        [InlineData(new string[] { "-1,4" }, "2,0", "3,3", 3, 3)]
+        [InlineData(new string[] { "4,-1" }, "2,2", "3,3", 3, 3)]
+        //[InlineData(new string[] { "4 4" }, "2,1", "4 3", 4, 3)] // ThrowOverflowOrFormatException
+        public void ReturnLocationString_GivenInvalidInputsFollowedByValidInput(IEnumerable<string> invalidInputs, string validInput, string universeInput, int gridWidth, int gridLength)
+        {
+            InputParser.ParseUniverse(universeInput);
+            _input.GetReadLine(invalidInputs);
+            _input.GetReadLine(validInput);
+            var expectedOutputLine1 = "Invalid input. Please try again.";
+            var expectedOutputLine2 = "Please enter the x, y coordinates for one live cell (e.g. '0,1').";
+
+            var result = _controller.CreateValidLocationString(gridWidth, gridLength);
+            var actualOutputLine1 = _output.GetWriteLine(0);
+            var actualOutputLine2 = _output.GetWriteLine(1);
+
+            Assert.Equal(validInput, result);
+            Assert.Equal(expectedOutputLine1, actualOutputLine1);
+            Assert.Equal(expectedOutputLine2, actualOutputLine2);
+        }
 
-        [Fact]
-        public void testName()
+        [Theory]
+        [InlineData(new string[] { "0,0" }, "-9,0", "0,0", -10, 3)] 
+        [InlineData(new string[] { "0,0" }, "-9,0", "0,0", 3, -10)] 
+        public void ThrowInvalidOperationException_GivenInvalidLocationInputsFollowedByValidLocationInputWithInvalidUniverseInput(IEnumerable<string> invalidInputs, string validInput, string universeString, int gridWidth, int gridLength)
         {
-            var universe = new Universe(3, 3);
-            Assert.Equal(3, universe.GridWidth);
+            InputParser.ParseUniverse(universeString);
+            _input.GetReadLine(invalidInputs);
+            _input.GetReadLine(validInput);
+    
+            Assert.Throws<InvalidOperationException>(() => _controller.CreateValidLocationString(gridWidth, gridLength));

```

Comment:
not sure what this is trying to test. you should never get to the `CreateValidLocationString` without a valid `Universe` already being parsed and stored in `_universe`